core/domain/classifier_services.py

```diff
@@ -127,6 +127,69 @@ def classify_string_classifier_rule(state, normalized_answer):
     return None
 
 
+
+
+def add_to_training_queue(exploration):
+    """Checks all states of the exploration for re-training conditions and
+    creates ClassifierTrainingJobModel instances in the datastore
+    accordingly.
+
+    Args:
+        exploration: Exploration. Domain object for an exploration.
+    """
+    states = exploration.states
+    for state_name in states:
+        state = states[state_name]
+        if not state.can_undergo_classification():
+            continue
+        exp_id = exploration.id
+        exp_version = exploration.version
+        training_data = state.get_training_data()
+        interaction_id = state.interaction.id
+        algorithm_id = feconf.INTERACTION_CLASSIFIER_MAPPING[
+            interaction_id]['algorithm_id']
+        classifier = get_classifier_from_exploration_attributes(
+            exp_id, exp_version, state_name)
+        if classifier:
```

Comment:
I am a little confused. Reading your code, the exp_version is apparently coming in from the exploration that is passed in, not the job. If this code is called immediately after the exploration is saved (as part of the save process), you should be able to pass in the correct version number. This code would then create the jobs with the right version number and then the StoreJobResult controller works as you describe.

What I am objecting to is that, in this code, it is not clear whether we are working with the exploration version after save, or before save, etc. My suggestion is to pass in the exp_id, necessary states, and version, and just create the job with the given exploration, states, and version (without any incrementing of version numbers etc.).